### PR TITLE
Create ReplicationController retry tooling 

### LINKED ISSUE #90881: Write CoreV1ReplicationController lifecycle test+promote - +7 coverage
# Progress `[8/8]`
* [x] Original approval issue: #88302 
* [x]  [APISnoop mock ticket](https://github.com/cncf/apisnoop/pull/289)
* [x]  test approval issue: [kubernetes/kubernetes#88588](https://github.com/kubernetes/kubernetes/issues/88588)
 * [x]  [update and improve](https://github.com/kubernetes/kubernetes/pull/ #90879)
* [x]  test pr : [#90880](https://github.com/kubernetes/kubernetes/pull/90880)
* [x]  two weeks soak start date: [testgrid-link](https://testgrid.k8s.io/google-gce#gci-gce-flaky&include-filter-by-regex=should%20test%20the%20lifecycle%20of%20a%20ReplicationController) Start 9 June 2020 [Master testgrid-link](https://testgrid.k8s.io/google-gce#gce-cos-master-flaky-repro&include-filter-by-regex=should%20test%20the%20lifecycle%20of%20a%20ReplicationController) 3 Flakes
* [x]  two weeks soak end date: [20 Oct 2020](https://testgrid.k8s.io/google-gce#gci-gce-flaky&include-filter-by-regex=should%20test%20the%20lifecycle%20of%20a%20ReplicationController) No Flakes for 2 weeks
* [X]  test promotion pr: [PR](https://github.com/kubernetes/kubernetes/pull/95713)

# Identify an untested feature Using APISnoop
According to this APIsnoop query, there are still some remaining ReplicationController endpoints which are untested.

```
SELECT
  operation_id,
  -- k8s_action,
  path,
  description
  FROM untested_stable_core_endpoints
  where path not like '%volume%'
  and operation_id ilike '%ReplicationController%'
 ORDER BY operation_id desc
 -- LIMIT 25
       ;
```

```
                     operation_id                      |                                path                                 |                          description                           
-------------------------------------------------------+---------------------------------------------------------------------+----------------------------------------------------------------
 replaceCoreV1NamespacedReplicationControllerStatus    | /api/v1/namespaces/{namespace}/replicationcontrollers/{name}/status | replace status of the specified ReplicationController
 readCoreV1NamespacedReplicationControllerStatus       | /api/v1/namespaces/{namespace}/replicationcontrollers/{name}/status | read status of the specified ReplicationController
 patchCoreV1NamespacedReplicationControllerStatus      | /api/v1/namespaces/{namespace}/replicationcontrollers/{name}/status | partially update status of the specified ReplicationController
 patchCoreV1NamespacedReplicationControllerScale       | /api/v1/namespaces/{namespace}/replicationcontrollers/{name}/scale  | partially update scale of the specified ReplicationController
 patchCoreV1NamespacedReplicationController            | /api/v1/namespaces/{namespace}/replicationcontrollers/{name}        | partially update the specified ReplicationController
 listCoreV1ReplicationControllerForAllNamespaces       | /api/v1/replicationcontrollers                                      | list or watch objects of kind ReplicationController
 deleteCoreV1CollectionNamespacedReplicationController | /api/v1/namespaces/{namespace}/replicationcontrollers               | delete collection of ReplicationController
(7 rows)
```

# Use API Reference to Lightly Document the Feature
* [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
* [client-go - ReplicationController](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/replicationcontroller.go)

# The mock test
## Test outline
1. Create a ReplicationController with a static label
2. Patch the ReplicationController with a new Label and updated data
3. Patch the ReplicationController's status
4. Patch/Scale the ReplicationController's Replica count to 2 Replicas
5. Get the ReplicationController
6. Replace the ReplicationController's Status
7. List all ReplicationControllers in all Namespaces find the ReplicationControllers(1) ensure that the ReplicationController is found and is patched
8. Delete Namespaced ReplicationControllers(1) via a Collection with a LabelSelector

## Example in Go
```go
package main

import (
  "encoding/json"
  "fmt"
  "flag"
  "os"

  v1 "k8s.io/api/core/v1"
  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
  "k8s.io/client-go/kubernetes"
  "k8s.io/apimachinery/pkg/runtime/schema"
  "k8s.io/apimachinery/pkg/types"
  autoscalingv1 "k8s.io/api/autoscaling/v1"
  "k8s.io/client-go/dynamic"
  "k8s.io/client-go/tools/clientcmd"
  //"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
)

func main() {
  // uses the current context in kubeconfig
  kubeconfig := flag.String("kubeconfig", fmt.Sprintf("%v/%v/%v", os.Getenv("HOME"), ".kube", "config"), "(optional) absolute path to the kubeconfig file")
  flag.Parse()
  config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
  if err != nil {
    fmt.Println(err)
  }
  // make our work easier to find in the audit_event queries
  config.UserAgent = "live-test-writing"
  // creates the clientset
  ClientSet, _ := kubernetes.NewForConfig(config)
  DynamicClientSet, _ := dynamic.NewForConfig(config)
  rcResource := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "replicationcontrollers"}

  testRcName := "rc-test"
  testRcNamespace := "default"
  testRcInitialReplicaCount := int32(1)
  testRcMaxReplicaCount := int32(2)

  rcTest := v1.ReplicationController{
      ObjectMeta: metav1.ObjectMeta{
          Name: testRcName,
          Labels: map[string]string{"test-rc-static": "true"},
      },
      Spec: v1.ReplicationControllerSpec{
          Replicas: &testRcInitialReplicaCount,
          Selector: map[string]string{"test-rc-static": "true"},
          Template: &v1.PodTemplateSpec{
              ObjectMeta: metav1.ObjectMeta{
                  Name: testRcName,
                  Labels: map[string]string{"test-rc-static": "true"},
              },
              Spec: v1.PodSpec{
                  Containers: []v1.Container{{
                      Name: testRcName,
                      Image: "nginx",
                  }},
              },
          },
      },
  }

  // Create a ReplicationController
  _, err = ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Create(&rcTest)
  if err != nil {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] created ReplicationController")

  // setup a watch for the RC
  rcWatch, err := ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Watch(metav1.ListOptions{LabelSelector: "test-rc-static=true"})
  if err != nil {
      fmt.Println(err)
      return
  }
  rcWatchChan := rcWatch.ResultChan()

  fmt.Println("[status] watching for all Replicas to be ready")
  for event := range rcWatchChan {
      rc, ok := event.Object.(*v1.ReplicationController)
      if ok != true {
          fmt.Println("Unable to fix type")
          return
      }
      if rc.Status.Replicas == testRcInitialReplicaCount && rc.Status.ReadyReplicas == testRcInitialReplicaCount {
          break
      }
  }

  rcLabelPatchPayload, err := json.Marshal(v1.ReplicationController{
      ObjectMeta: metav1.ObjectMeta{
          Labels: map[string]string{"test-rc": "patched"},
      },
  })
  if err != nil {
    fmt.Println(err)
    return
  }
  // Patch the ReplicationController
  _, err = ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Patch(testRcName, types.StrategicMergePatchType, []byte(rcLabelPatchPayload))
  if err != nil {
    fmt.Println(err)
    return
  }
  fmt.Println("[status] patched ReplicationController")

  rcStatusPatchPayload, err := json.Marshal(map[string]interface{}{
    "status": map[string]interface{}{
      "readyReplicas": 0,
      "availableReplicas": 0,
    },
  })
  if err != nil {
      fmt.Println(err)
      return
  }

  // Patch the ReplicationController's status
  rcStatus, err := ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Patch(testRcName, types.StrategicMergePatchType, []byte(rcStatusPatchPayload), "status")
  if err != nil {
      fmt.Println(err)
      return
  }
  if rcStatus.Status.ReadyReplicas != 0 {
      fmt.Println("[error] failed to patch ReplicationController status; ReadyReplicas != 0")
      return
  }
  fmt.Println("[status] patched status of ReplicationController")

  rcStatusUnstructured, err := DynamicClientSet.Resource(rcResource).Namespace(testRcNamespace).Get(testRcName, metav1.GetOptions{}, "status")
  if err != nil {
      fmt.Println(err)
      return
  }
  rcStatusUjson, _ := json.Marshal(rcStatusUnstructured)
	json.Unmarshal(rcStatusUjson, &rcStatus)
  if rcStatus.Status.Replicas != testRcInitialReplicaCount {
     fmt.Println("[error] Replica count does not match initial Replica count")
  }
  fmt.Println("[status] fetched ReplicationController status")

  rcScalePatchPayload, err := json.Marshal(autoscalingv1.Scale{
      Spec: autoscalingv1.ScaleSpec{
          Replicas: 2,
      },
  })
  if err != nil {
      fmt.Println(err)
      return
  }

  // Patch the ReplicationController's scale
  rcScale, err := ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Patch(testRcName, types.StrategicMergePatchType, []byte(rcScalePatchPayload), "scale")
  if err != nil {
      fmt.Println(err)
      return
  }
  rcScaleReplicasExpect := int32(2)
  if rcScale.Status.Replicas == rcScaleReplicasExpect {
      fmt.Println("[error] failed to patch ReplicationController scale; ReadyReplicas != 2")
      return
  }
  fmt.Println("[status] patched scale of ReplicationController")

  fmt.Println("[status] watching for all Replicas to be ready")
  for event := range rcWatchChan {
      rc, ok := event.Object.(*v1.ReplicationController)
      if ok != true {
          fmt.Println("Unable to fix type")
          return
      }
      if rc.Status.Replicas == testRcMaxReplicaCount && rc.Status.ReadyReplicas == testRcMaxReplicaCount {
          break
      }
  }

  // Get the ReplicationController
  rc, err := ClientSet.CoreV1().ReplicationControllers(testRcNamespace).Get(testRcName, metav1.GetOptions{})
  if err != nil {
      fmt.Println(err)
      return
  }
  if rc.ObjectMeta.Labels["test-rc"] != "patched" {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] fetched ReplicationController")

  rcStatusUpdatePayload := rc
  rcStatusUpdatePayload.Status.AvailableReplicas = 1
  rcStatusUpdatePayload.Status.ReadyReplicas = 1

  // Replace the ReplicationController's status	
  rcStatus, err = ClientSet.CoreV1().ReplicationControllers(testRcNamespace).UpdateStatus(rcStatusUpdatePayload)
  if err != nil {
      fmt.Println(err)
      return
  }
  if rcStatus.Status.ReadyReplicas != 1 {
      fmt.Println("[error] failed to patch ReplicationController status; ReadyReplicas != 1")
      return
  }
  fmt.Println("[status] updated ReplicationController status")

  fmt.Println("[status] watching for all Replicas to be ready")
  for event := range rcWatchChan {
      rc, ok := event.Object.(*v1.ReplicationController)
      if ok != true {
          fmt.Println("Unable to fix type")
          return
      }
      if rc.Status.Replicas == testRcMaxReplicaCount && rc.Status.ReadyReplicas == testRcMaxReplicaCount {
          fmt.Println("[status] all Replicas are ready")
          break
      }
  }

  rcs, err := ClientSet.CoreV1().ReplicationControllers("").List(metav1.ListOptions{LabelSelector: "test-rc-static=true"})
  if err != nil {
      fmt.Println(err)
      return
  }
  if len(rcs.Items) == 0 {
      fmt.Println("[error] no ReplicationController were found")
  }
  foundRc := false
  for _, rcItem := range rcs.Items {
      if rcItem.ObjectMeta.Name == testRcName &&
         rcItem.ObjectMeta.Namespace == testRcNamespace &&
         rcItem.ObjectMeta.Labels["test-rc-static"] == "true" &&
         rcItem.ObjectMeta.Labels["test-rc"] == "patched" && 
         rcItem.Status.Replicas == testRcMaxReplicaCount &&
         rcItem.Status.ReadyReplicas == testRcMaxReplicaCount {
         foundRc = true
      }
  }
  if foundRc == false {
      fmt.Println("[error] unable to find ReplicationController")
      return
  }
  fmt.Println("[status] retrieved all ReplicationControllers selecting with LabelSelector")

  // Delete ReplicationController
  err = ClientSet.CoreV1().ReplicationControllers(testRcNamespace).DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: "test-rc-static=true"})
  if err != nil {
    fmt.Println(err)
    return
  }
  fmt.Println("[status] deleted ReplicationController")

  fmt.Println("[status] complete")

}
```

```go
[status] created ReplicationController
[status] watching for all Replicas to be ready
[status] patched ReplicationController
[status] patched status of ReplicationController
[status] fetched ReplicationController status
[status] patched scale of ReplicationController
[status] watching for all Replicas to be ready
[status] fetched ReplicationController
[status] updated ReplicationController status
[status] watching for all Replicas to be ready
[status] all Replicas are ready
[status] retrieved all ReplicationControllers selecting with LabelSelector
[status] deleted ReplicationController
[status] complete
```

# Verify Increase it Coverage with APISnoop
Discover useragents:

```
select distinct useragent from audit_event where bucket='apisnoop' and useragent not like 'kube%' and useragent not like 'coredns%' and useragent not like 'kindnetd%' and useragent like 'live%';
```

```
     useragent     
-------------------
 live-test-writing
(1 row)
```

See all endpoints hit from this test

```
select * from endpoints_hit_by_new_test where useragent like 'live%'; 
```

```
     useragent     |                     operation_id                      | hit_by_ete | hit_by_new_test 
-------------------+-------------------------------------------------------+------------+-----------------
 live-test-writing | createCoreV1NamespacedReplicationController           | t          |               2
 live-test-writing | deleteCoreV1CollectionNamespacedReplicationController | f          |               2
 live-test-writing | listCoreV1NamespacedReplicationController             | t          |               1
 live-test-writing | listCoreV1ReplicationControllerForAllNamespaces       | f          |               1
 live-test-writing | patchCoreV1NamespacedReplicationController            | f          |               2
 live-test-writing | patchCoreV1NamespacedReplicationControllerScale       | f          |               2
 live-test-writing | patchCoreV1NamespacedReplicationControllerStatus      | f          |               2
 live-test-writing | readCoreV1NamespacedReplicationController             | t          |               1
 live-test-writing | readCoreV1NamespacedReplicationControllerStatus       | f          |               1
 live-test-writing | replaceCoreV1NamespacedReplicationControllerStatus    | f          |               2
(10 rows)
```

Display test coverage change

```
select * from projected_change_in_coverage;
```

```
   category    | total_endpoints | old_coverage | new_coverage | change_in_number 
---------------+-----------------+--------------+--------------+------------------
 test_coverage |             438 |          190 |          197 |                7
(1 row)
```

# Final notes
If a test with these calls gets merged, ****test coverage will go up by 7 points****

This test is also created with the goal of conformance promotion.

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup
Replaces: #90629
Fixes: #89746 
Fixes: #90881
**What this PR does / why we need it**:
Adds useful ReplicationController patch functions for ensuring that ReplicationControllers get patched

**Which issue(s) this PR fixes**:
[#89746 (comment)](https://github.com/kubernetes/kubernetes/pull/89746#issuecomment-610112234)

**Special notes for your reviewer**:
This is needed for #89746

**Does this PR introduce a user-facing change?**:

```
NONE
```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```
NONE
```

/sig testing

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
/ok-to-test
/priority important-longterm

File: test/e2e/apps/rc.go [hunk 1]
```diff
@@ -558,3 +563,56 @@ func updateReplicationControllerWithRetries(c clientset.Interface, namespace, na
...
 	return rc, pollErr
 }
+
+type rcPatchFunc func(rc *v1.ReplicationController) ([]byte, error)
+
+func patchRCOrDie(cs kubernetes.Interface, dc dynamic.Interface, ns string, name string, f rcPatchFunc, subresources ...string) (updated *v1.ReplicationController) {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		old := getRCStatusOrDie(dc, ns, name)
+		patchBytes, err := f(old)
+		framework.ExpectNoError(err, "Failed to run ReplicationController patch function for resource %s", name)
```
Comment (neolit123) [inline test/e2e/apps/rc.go:573]: please make all error annotations to start lowercase
`the ReplicationController....`
[1]

File: test/e2e/apps/rc.go [hunk 2]
```diff
@@ -558,3 +563,56 @@ func updateReplicationControllerWithRetries(c clientset.Interface, namespace, na
...
+		rc, err := cs.CoreV1().ReplicationControllers(ns).Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if rc.Status.ObservedGeneration < rc.Generation {
+			return false, fmt.Errorf("the ObservedGeneration for ReplicationController %s in namespaced %s is %v", name, ns, rc.Status.ObservedGeneration)
+		}
+		return true, nil
+	})
+	framework.ExpectNoError(err, "The ReplicationController %q in namespace %q did not update after %v", name, ns, schedulingTimeout)
```
Comment (neolit123) [inline test/e2e/apps/rc.go:597]: [1]

File: test/e2e/apps/rc.go [hunk 3]
```diff
@@ -558,3 +563,56 @@ func updateReplicationControllerWithRetries(c clientset.Interface, namespace, na
...
+}
+
+func getRCStatusOrDie(dc dynamic.Interface, ns string, name string) *v1.ReplicationController {
+	rcStatusResource := v1.SchemeGroupVersion.WithResource("replicationcontroller")
+	unstruct, err := dc.Resource(rcStatusResource).Namespace(ns).Get(context.TODO(), name, metav1.GetOptions{}, "status")
+	rc, err := unstructuredToRC(unstruct)
+	framework.ExpectNoError(err, "Failed to get ReplicationController %q in namespace %q", name, ns)
```
Comment (neolit123) [inline test/e2e/apps/rc.go:604]: [1]

**Comment by neolit123** [comment]
added a couple of minor comments, other than that LGTM.
ideally the next PR should try to utilize the functions so that we can amend them if needed.

File: test/e2e/apps/rc.go [hunk 4]
```diff
@@ -558,3 +563,56 @@ func updateReplicationControllerWithRetries(c clientset.Interface, namespace, na
...
+			return false, fmt.Errorf("the ObservedGeneration for ReplicationController %s in namespaced %s is %v", name, ns, rc.Status.ObservedGeneration)
+		}
+		return true, nil
+	})
+	framework.ExpectNoError(err, "The ReplicationController %q in namespace %q did not update after %v", name, ns, schedulingTimeout)
+}
+
+func getRCStatusOrDie(dc dynamic.Interface, ns string, name string) *v1.ReplicationController {
+	rcStatusResource := v1.SchemeGroupVersion.WithResource("replicationcontroller")
+	unstruct, err := dc.Resource(rcStatusResource).Namespace(ns).Get(context.TODO(), name, metav1.GetOptions{}, "status")
```
Comment (oomichi) [inline test/e2e/apps/rc.go:602]: Need to add `framework.ExpectNoError(err)` here

**Comment by BobyMCbobs** [comment]
This may not be needed if the [watch event tooling](https://github.com/kubernetes/kubernetes/pull/91416) fixes the patching of the scale

File: test/e2e/apps/rc.go [hunk 5]
```diff
@@ -558,3 +563,56 @@ func updateReplicationControllerWithRetries(c clientset.Interface, namespace, na
...
+
+func waitForRCToBeProcessed(cs kubernetes.Interface, ns string, name string) {
+	ginkgo.By("Waiting for the rc to be processed")
+	err := wait.PollImmediate(framework.Poll, schedulingTimeout, func() (bool, error) {
+		rc, err := cs.CoreV1().ReplicationControllers(ns).Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if rc.Status.ObservedGeneration < rc.Generation {
+			return false, fmt.Errorf("the ObservedGeneration for ReplicationController %s in namespaced %s is %v", name, ns, rc.Status.ObservedGeneration)
```
Comment (oomichi) [inline test/e2e/apps/rc.go:593]: It is better to ouput `rc.Generation` into the error message to investigate issues when this error happens.

**Review by oomichi — COMMENTED** [review]
One more cooment

**Pull request closed without merging**